File: code/java/ch10/practice/Practice05GenerateSubsets.java
package ch10.practice;

import java.util.*;

/**
 * Practice 05: Generate All Subsets
 * ==============================
 * Chapter 10: The Magic of Recursion
 *
 * PROBLEM: Given an array of distinct integers, return all possible subsets.
 *          The result should be sorted: first by subset size, then lexicographically.
 *
 * ALGORITHM: Backtracking — at each element, include it or exclude it.
 *
 * EXAMPLES:
 *   solve(new int[]{})      = [[]]
 *   solve(new int[]{1})     = [[], [1]]
 *   solve(new int[]{1,2,3}) = [[], [1], [2], [3], [1,2], [1,3], [2,3], [1,2,3]]
 *
 * CONSTRAINTS: 0 <= nums.length <= 15, all elements are distinct
 *
 * INSTRUCTIONS: Replace the body of solve() with your solution.
 */
public class Practice05GenerateSubsets {
    public static List<List<Integer>> solve(int[] nums) {
        // TODO: Replace this with your solution
        return new ArrayList<>();
    }

    // ── Do not change anything below this line ──────────────────────
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String line = sc.nextLine().trim();
        int[] nums;
        if (line.isEmpty()) {
            nums = new int[0];
        } else {
            String[] parts = line.split("\\s+");
            nums = new int[parts.length];
            for (int i = 0; i < parts.length; i++) nums[i] = Integer.parseInt(parts[i]);
        }
        List<List<Integer>> result = solve(nums);
        for (List<Integer> subset : result) {
            System.out.println(subset);
        }
        sc.close();
    }
}

File: code/cpp/ch10/practice/practice_05_generate_subsets.cpp
/*
 * Practice 5: Generate All Subsets
 * ==================================
 * Chapter 10: The Magic of Recursion
 *
 * PROBLEM:
 *   Given an array of distinct integers, return all possible subsets
 *   (the power set). The result should be sorted: each subset sorted
 *   internally, and subsets sorted lexicographically.
 *
 * EXAMPLES:
 *   solve({})      -> {{}}
 *   solve({1})     -> {{}, {1}}
 *   solve({1,2,3}) -> {{}, {1}, {1,2}, {1,2,3}, {1,3}, {2}, {2,3}, {3}}
 *
 * CONSTRAINTS:
 *   0 <= nums.size() <= 10
 *   All elements are distinct.
 *
 * INSTRUCTIONS:
 *   Replace the body with your recursive/backtracking solution.
 */

#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

vector<vector<int>> solve(vector<int> nums) {
    // TODO: Replace this with your solution
    return {};
}

// -- Do not change anything below this line --------------------------
int main() {
    int n;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++) cin >> nums[i];
    vector<vector<int>> result = solve(nums);
    cout << "[";
    for (int i = 0; i < (int)result.size(); i++) {
        if (i > 0) cout << ", ";
        cout << "[";
        for (int j = 0; j < (int)result[i].size(); j++) {
            if (j > 0) cout << ", ";
            cout << result[i][j];
        }
        cout << "]";
    }
    cout << "]" << endl;
    return 0;
}
